### PR TITLE
feat(notifications): use pagination on the backend layer

### PR DESCRIPTION
**Pull request by mareklibra**
## Hey, I just made a Pull Request!

The NotificationsPage uses pagination by the backend to avoid large datasets to be loaded into frontend.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

## Depends On
This PR will be rebased after:
- [x] https://github.com/backstage/backstage/pull/23173

### CODE REVIEW

File: plugins/notifications-backend/src/service/router.ts [hunk 1]
```diff
@@ -205,15 +205,19 @@ export async function createRouter(
...
     if (req.query.created_after) {
-      const sinceEpoch = Date.parse(req.query.created_after.toString());
+      const sinceEpoch = Date.parse(String(req.query.created_after));
       if (isNaN(sinceEpoch)) {
         throw new InputError('Unexpected date format');
       }
       opts.createdAfter = new Date(sinceEpoch);
     }
 
     const notifications = await store.getNotifications(opts);
```
Comment (drodil) [inline plugins/notifications-backend/src/service/router.ts:215]: Could do this in `const [notifications, totalCount] = await Promise.all([store.getNotifications(opts), store.getNotificationsCount(opts)]);

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts [hunk 2]
```diff
@@ -165,6 +164,17 @@ export class DatabaseNotificationsStore implements NotificationsStore {
...
   }
 
+  async getNotificationsCount(options: NotificationGetOptions) {
+    const countOptions: NotificationGetOptions = { ...options };
+    countOptions.limit = undefined;
+    countOptions.offset = undefined;
+    countOptions.sort = null;
+    const notificationQuery = this.getNotificationsBaseQuery(countOptions);
+    const response = await notificationQuery.count('* as CNT');
+    const totalCount = Number.parseInt(response[0].CNT.toString(), 10);
```
Comment (drodil) [inline plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts:174]: You could do just `return Number(response[0].CNT)`

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.test.ts [hunk 3]
```diff
@@ -215,6 +215,87 @@ describe.each(databases.eachSupportedId())(
...
+        const id2 = uuid();
+        const id3 = uuid();
+        const id4 = uuid();
+        const id5 = uuid();
+        const id6 = uuid();
+        const id7 = uuid();
+        await insertNotification({
+          id: id1,
+          ...testNotification,
+          created: new Date(Date.now() - 1 * 60 * 60 * 1000 /* an hour ago */),
```
Comment (drodil) [inline plugins/notifications-backend/src/database/DatabaseNotificationsStore.test.ts:231]: Should you use `now` instead `Date.now()`, not big difference but all other are using that

Comment (mareklibra) [reply plugins/notifications-backend/src/database/DatabaseNotificationsStore.test.ts]: done

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts (hunk 2 above)
Comment (mareklibra) [reply plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts]: done

File: plugins/notifications-backend/src/service/router.ts (hunk 1 above)
Comment (mareklibra) [reply plugins/notifications-backend/src/service/router.ts]: done

**Review by Rugvip — APPROVED** [review]
:shipit:

**Comment by mareklibra** [comment]
Debug in progress...
It seems that `query.orderBy('created', 'desc');` does not work on mysql but works on other engines, including sqlite

**Pull request merged**